inserting events at positions based on their timestamps

pfichtner opened the pull request:
Based on the comment regarding performance, here's another approach to investigate: 
Instead of having sorted streams when accessing the list, insert the events based on there timestamp at the right index when adding them to the list. Of course the more often the list is read/accessed the more performance you will gain. I guess that performance will increase even if it's accessed just once because creating a sorted stream is a relatively expensive operation since the underlying collection has to be sorted in memory before the first element can be accessed. 

However, there are some semantic changes due to this change: There have been several statemens that blindly accesses head/tail elements in the list which is on `master` the sequence the elements were added and on `sortedlist` the first or last element based on the events' timestamps. However: All test are green so it seems that this already was the expected behavior. 
Could you please verify using some big gc logs since I only can do some repeating tests with some small gc logs can of course lead to completely different measurements.

mgm3746 commented:
I ran against a 50g gc log and got the following execution times:

With PR: 10 mins, 52 secs
Without PR: 10 mins, 54 secs

I'm not aware of any good use case for a giant gc log file. Users should be rotating them in some way. And even if they are not, I would split out the relevant time period when doing analysis.

So I don't think performance on large gc files is a priority, and we could do whatever makes sense with the order of the gc logging. 

The reason there is a sort option is because a customer was using splunk to aggregate gc logging, and their process messed up the order of the gc logging. 

I think sometimes the JVM gets so stressed out that it logs gc events in the wrong order. At least with JDK8. And when that happens, I just reorder it. It's good to know though, for analysis, whether the reordering happened or not, so can we have an event that notes when logging is found out of order? Or, continue with the TimeWarpException "Bad order" and have the --reorder flag suppress that exeption?

I haven't seen this so much lately, but it used to be that sometimes we would see gc logging that spanned multiple JVM runs.  That would be another reason to keep the TimeWarpException "Bad order"  and only override it consciously.

mgm3746 commented:
It looks like all the exceptions are in place. If you are happy with this, I will merge it in. Let me know.

pfichtner commented:
I made another commit what should gain some performance: I check if the element that should been added has a greater timestamp than the head element. 
After reading your comment I refactored the method so that throwing Exceptions can be added easily without changing the code again and added a comment for that. So please merge if applicable.

mgm3746 commented:
Looks great, thank you, Peter!
